python/ray/air/_internal/torch_utils.py

```diff
@@ -1,71 +1,31 @@
 
 import numpy as np
 import pandas as pd
-import torch
 
-import ray
+from ray.air._internal.accelerator_utils import try_import_torch
 from ray.air.util.data_batch_conversion import _unwrap_ndarray_object_type_if_needed
 
-from ray._private.accelerators.hpu import HPU_PACKAGE_AVAILABLE
 
-if HPU_PACKAGE_AVAILABLE:
-    import habana_frameworks.torch.hpu as torch_hpu
+torch, _ = try_import_torch()
 
 
 def get_devices() -> List[torch.device]:
     """Gets the correct torch device list configured for this process.
 
-    Returns a list of torch CUDA devices allocated for the current worker.
-    If no GPUs are assigned, then it returns a list with a single CPU device.
-
-    Assumes that `CUDA_VISIBLE_DEVICES` is set and is a
-    superset of the `ray.get_gpu_ids()`.
+    Returns a list of torch Accelerator devices allocated for the current worker.
+    If no Accelerators are assigned, then it returns a list with a single CPU device.
     """
-    if torch.cuda.is_available():
-        # GPU IDs are assigned by Ray after you specify "use_gpu"
-        # GPU `ray.get_gpu_ids()` may return ints or may return strings.
-        # We should always convert to strings.
-        gpu_ids = [str(id) for id in ray.get_gpu_ids()]
-
-        device_ids = []
-
-        if len(gpu_ids) > 0:
-            cuda_visible_str = os.environ.get("CUDA_VISIBLE_DEVICES", "")
-            if cuda_visible_str and cuda_visible_str != "NoDevFiles":
-                cuda_visible_list = cuda_visible_str.split(",")
-            else:
-                cuda_visible_list = []
-
-            # By default, there should only be one GPU ID if `use_gpu=True`.
-            # If there are multiple GPUs, return a list of devices.
-            # If using fractional GPUs, these IDs are not guaranteed
-            # to be unique across different processes.
-            for gpu_id in gpu_ids:
-                try:
-                    device_ids.append(cuda_visible_list.index(gpu_id))
-                except IndexError:
-                    raise RuntimeError(
-                        "CUDA_VISIBLE_DEVICES set incorrectly. "
-                        f"Got {cuda_visible_str}, expected to include {gpu_id}. "
-                        "Did you override the `CUDA_VISIBLE_DEVICES` environment"
-                        " variable? If not, please help file an issue on Github."
-                    )
-
-        else:
-            # If called on the driver or outside of Ray Train, return the
-            # 0th device.
-            device_ids.append(0)
+    from ray.air._internal.accelerator_utils import (
+        get_torch_device_manager_by_runtime_context,
+    )
 
-        devices = [torch.device(f"cuda:{device_id}") for device_id in device_ids]
-    elif HPU_PACKAGE_AVAILABLE and torch_hpu.is_available():
-        devices = [torch.device("hpu")]
+    torch_device_manager = get_torch_device_manager_by_runtime_context()
+    if torch_device_manager:
```

Comment:
Thinking out loud. We should consider adding `cpu` as one of the accelerator type, and always have a device manager instead of having this fallback logic here and there.

Reply:
That's correct. However, we recommend using CUDADeviceManager as the default value so that we can use `device_module.is_available()` to determine device availability. When the device is not available, it can fall back to using the CPU (which is considered the host rather than a device).